### PR TITLE
Store source, not sname, in Symbol

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91058
* __->__ #91057

I'm going to need this in the follow up PR. Instead of storing only Source.name() in Symbol, I now store a full on Source. Lots of replumbing reoccurs. In particular:

- Move Source to torch._guards to break cycles
- I have to add TensorPropertySource and NegateSource to handle x.size()[0] and -x codegen that I was doing with string manipulation previously
- I tighten up invariants so that I never pass source=None; instead I pass ConstantSource (these are constant sources right) and test for that rather than source being missing. I think this is more parsimonious
- Some mypy wobbles from new imports

I didn't move LocalSource and friends to torch._guards, but I ended up needing to access them in a few places. The main annoyance with moving these is that then I also need to move the bytecode codegen stuff, and that's not so easy to move without bringing in the kitchen sink.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -935,29 +933,30 @@ def _clone_input(value):
...
     if type(e) in (torch.Tensor, torch.nn.Parameter) or (
         ignore_subclass and isinstance(e, torch.Tensor)
     ):
-        static_shapes = static_shapes or config.dynamic_shapes is False
-        if type(e) is torch.nn.Parameter:
-            # Always static for params
-            static_shapes = True
+        static_shapes = (
+            source is None
+            or type(e) is torch.nn.Parameter
```
Comment (albanD) [inline torch/_dynamo/variables/builder.py:947]: Why is and not isinstance here?

**Review by albanD — APPROVED** [review]
Sounds good

File: torch/_guards.py [hunk 2]
```diff
@@ -343,3 +343,27 @@ def tracing(context: TracingContext):
...
+@dataclasses.dataclass
+class Source:
+    def reconstruct(self, codegen):
+        raise NotImplementedError()
+
+    def guard_source(self):
+        raise NotImplementedError()
+
+    def name(self):
+        raise NotImplementedError()
```
Comment (voznesenskym) [inline torch/_guards.py:358]: Feels like an ABC?

File: torch/_guards.py [hunk 3]
```diff
@@ -343,3 +343,27 @@ def tracing(context: TracingContext):
...
+
+    def make_guard(self, fn, is_volatile=False):
+        if self.guard_source() is GuardSource.CONSTANT:
+            raise NotImplementedError()
```
Comment (voznesenskym) [inline torch/_guards.py:362]: Is this correct?

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -207,10 +207,10 @@ def __repr__(self):
 
     # These methods are metaprogrammed in below
     def sym_int(self) -> "SymNode":
-        ...
+        raise AssertionError("should have been overridden")
 
     def sym_float(self) -> "SymNode":
-        ...
+        raise AssertionError("should have been overridden")
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:213]: nit: Feel like a notimplemented?

File: torch/_subclasses/meta_utils.py [hunk 5]
```diff
@@ -159,10 +160,13 @@ def meta_storage(self, s, callback):
...
-        self, t, shape_env=None, callback=lambda t: t(), sname: Optional[str] = None
+        self, t, shape_env=None, callback=lambda t: t(), source: Optional[Source] = None
     ):
-        if sname is None:
-            sname = f"__unknown_tensor{len(self.tensor_memo)}"
+        if source is None:
+            from torch._dynamo.source import ConstantSource
+
+            # TODO: make a dedicated UnknownSource for this?
+            source = ConstantSource(f"__unknown_tensor{len(self.tensor_memo)}")
```
Comment (voznesenskym) [inline torch/_subclasses/meta_utils.py:169]: Agreed

**Review by voznesenskym — APPROVED** [review]

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:947]: If it is a Parameter subclass then we need to preserve the particular subclass, but now we have no guarantees that the subclass isn't doing funny business on operations. With Parameter we know what the semantics are. (Also, this is a preexisting condition.)

File: torch/_guards.py (hunk 2 above)
Comment (ezyang) [reply torch/_guards.py:358]: Sure, we can do that in a follow up ;)

File: torch/_guards.py (hunk 3 above)
Comment (ezyang) [reply torch/_guards.py:362]: I mean, it's your code!! lol. I would hope you know.

The constant source stuff is pretty mushy for me. But it at least sort of makes sense that you shouldn't ever be guarding on them; they're constant after all, why would you guard on something that the user has said should never change from iteration to iteration.

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:213]: no, it's not that it's not implemented, this is just a stub that will get overwritten by metaprogramming later. If you ever actually reach this code that means something wrong happened in our code, therefore assert.

**Comment by zou3519** [comment]
Sorry this was my fault. The problem is that in the internal build, functorch is set up differently from the OSS build, so it is not always included with PyTorch.

The short-term fix is simple, change https://github.com/pytorch/pytorch/blob/6e0cd8b91ed8cb45444224bc24552956016338ec/torch/_dynamo/variables/builder.py#L14 to from torch._ops import PyOperator.

In the medium term I am going to figure out how to make the internal build align better with the OSS build

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**